### ISSUE TITLE
The accuracy staying 0.500+-0.000 all the time when using MobileNet 

### ISSUE THREAD

**Issue by VincentVon** [davidsandberg/facenet#672]
Hello @davidsandberg , 
Thanks to your excellent work at first, and i have learned a lot in your project.
But I came up an issue about using new network: MobileNet.
The training process seems all good, but the validation process on LFW dataset is not correct.
The accuracy staying 0.5000+-0.000 after every epoch, but the training loss keeps falling.
I spent TWO days focus on this issue, and I found the embeddings at validation process are almost same.
I think that is why the accuracy staying 0.5000.
When i change phase_train_placeholder: False to True in evaluate function, the accuracy is not 0.500 and seems normal, but we all know phase_train_placeholder should be False in evaluate function.
I cannot deal with this issue, and any advice could be helpful. BTW, my GPU is GTX1070, 8G.
Here is MobileNet code, changing from @Zehaos code.

`

	from __future__ import absolute_import
	from __future__ import division
	from __future__ import print_function
	import tensorflow as tf
	import tensorflow.contrib.slim as slim

	def mobilenet(inputs, is_training=True, dropout_keep_prob=0.8, bottleneck_layer_size=128, reuse=None, scope='MobileNet'):
		""" MobileNet
		More detail, please refer to Google's paper(https://arxiv.org/abs/1704.04861).
		Args:
		  inputs: a tensor of size [batch_size, height, width, channels].
		  num_classes: number of predicted classes.
		  is_training: whether or not the model is being trained.
		  scope: Optional scope for the variables.
		Returns:
		  logits: the pre-softmax activations, a tensor of size
			[batch_size, `num_classes`]
		  end_points: a dictionary from components of the network to the corresponding
			activation.
		"""
		width_multiplier = 1

		def _depthwise_separable_conv(inputs, num_pwc_filters, width_multiplier, sc, downsample=False):
			""" Helper function to build the depth-wise separable convolution layer.
			"""
			num_pwc_filters = round(num_pwc_filters * width_multiplier)
			_stride = 2 if downsample else 1

			# skip pointwise by setting num_outputs=None
			depthwise_conv = slim.separable_convolution2d(inputs, num_outputs=None, stride=_stride, depth_multiplier=1, kernel_size=[3, 3], scope=sc + '/depthwise_conv')

			bn = slim.batch_norm(depthwise_conv, scope=sc + '/dw_batch_norm')
			pointwise_conv = slim.convolution2d(bn, num_pwc_filters, kernel_size=[1, 1], scope=sc + '/pointwise_conv')
			bn = slim.batch_norm(pointwise_conv, scope=sc + '/pw_batch_norm')
			return bn

		with tf.variable_scope(scope) as sc:
			end_points_collection = sc.name + '_end_points'
			with slim.arg_scope([slim.convolution2d, slim.separable_convolution2d], activation_fn=None, outputs_collections=[end_points_collection]):
				with slim.arg_scope([slim.batch_norm], is_training=is_training, activation_fn=tf.nn.relu, fused=True):
					net = slim.convolution2d(inputs, round(32 * width_multiplier), [3, 3], stride=2, padding='SAME', scope='conv_1')
					net = slim.batch_norm(net, scope='conv_1/batch_norm')
					net = _depthwise_separable_conv(net, 64, width_multiplier, sc='conv_ds_2')
					net = _depthwise_separable_conv(net, 128, width_multiplier, downsample=True, sc='conv_ds_3')
					net = _depthwise_separable_conv(net, 128, width_multiplier, sc='conv_ds_4')
					net = _depthwise_separable_conv(net, 256, width_multiplier, downsample=True, sc='conv_ds_5')
					net = _depthwise_separable_conv(net, 256, width_multiplier, sc='conv_ds_6')
					net = _depthwise_separable_conv(net, 512, width_multiplier, downsample=True, sc='conv_ds_7')

					net = _depthwise_separable_conv(net, 512, width_multiplier, sc='conv_ds_8')
					net = _depthwise_separable_conv(net, 512, width_multiplier, sc='conv_ds_9')
					net = _depthwise_separable_conv(net, 512, width_multiplier, sc='conv_ds_10')
					net = _depthwise_separable_conv(net, 512, width_multiplier, sc='conv_ds_11')
					net = _depthwise_separable_conv(net, 512, width_multiplier, sc='conv_ds_12')

					net = _depthwise_separable_conv(net, 1024, width_multiplier, downsample=True, sc='conv_ds_13')
					net = _depthwise_separable_conv(net, 1024, width_multiplier, sc='conv_ds_14')
					net = slim.avg_pool2d(net, [7, 7], scope='avg_pool_15')

			end_points = slim.utils.convert_collection_to_dict(end_points_collection)
			net = tf.squeeze(net, [1, 2], name='SpatialSqueeze')
			end_points['squeeze'] = net
			logits = slim.fully_connected(net, bottleneck_layer_size, activation_fn=None, scope='fc_16')
			predictions = slim.softmax(logits, scope='Predictions')

			end_points['Logits'] = logits
			end_points['Predictions'] = predictions

		return logits, end_points

	def inference(images, keep_probability, phase_train=True, bottleneck_layer_size=128, weight_decay=0.0, reuse=None):
		batch_norm_params = {
			# Decay for the moving averages.
			'decay': 0.995,
			# epsilon to prevent 0s in variance.
			'epsilon': 0.001,
			# force in-place updates of mean and variance estimates
			'updates_collections': None,
			# Moving averages ends up in the trainable variables collection
			'variables_collections': [tf.GraphKeys.TRAINABLE_VARIABLES],
		}

		with slim.arg_scope([slim.convolution2d, slim.separable_convolution2d],
						 weights_initializer=tf.truncated_normal_initializer(stddev=0.1),
						 weights_regularizer=slim.l2_regularizer(weight_decay),
						 biases_initializer=slim.init_ops.zeros_initializer(),
						 normalizer_fn=slim.batch_norm,
						 normalizer_params=batch_norm_params):
			return mobilenet(images,
						    is_training=phase_train,
					            dropout_keep_prob=keep_probability,
						    bottleneck_layer_size=bottleneck_layer_size,
						    reuse=reuse)
`

And parameters are 
`

	logs_base_dir = '...'
	models_base_dir = '...'
	gpu_memory_fraction = 1.0
	pretrained_model = False
	data_dir = '...'
	model_def = 'src.models.mobilenet_v1'
	max_nrof_epochs = 100
	batch_size = 32
	image_size = 224
	epoch_size = 10

	embedding_size = 256
	random_crop = True
	random_flip = True
	random_rotate = True

	keep_probability = 0.8
	weight_decay = 5e-5
	center_loss_factor = 1e-2
	center_loss_alfa = 0.9
	learning_rate = -1

	optimizer = 'ADAGRAD'
	learning_rate_decay_epochs = 100
	learning_rate_decay_factor = 1.0
	moving_average_decay = 0.9999
	seed = 666
	nrof_preprocess_threads = 4
	log_histograms = True
	learning_rate_schedule_file = '...'
	filter_filename = ''
	filter_percentile = 100.0
	filter_min_nrof_images_per_class = 0

	# Parameters for validation on LFW
	lfw_pairs = '...'
	lfw_file_ext = 'png'
	lfw_dir = '...'
	lfw_batch_size = 100
	lfw_nrof_folds = 10
`

Thanks again.

**Comment by Julyustc** [comment]
Hello, are you solving this problem now? I have the same problem, if it is resolved, can you share the solution about that?Thank you very much!

**Comment by VincentVon** [comment]
@Julyustc , hi
I didn't solve this problem yet, it is really confusing.
I checked the feature map or embeddings after every layer (you can check these in endpoints), and I found the feature vectors about every image are approaching same with layers are deeper.
I still working on it, and I decide to check layer weights.
Keep contact, alright?

**Comment by hgffly** [comment]
hi @VincentVon, @Julyustc  , try this, I met the same problems before but I have resolved it, hope it's helpful
with tf.control_dependencies(tf.get_collection(tf.GraphKeys.UPDATE_OPS)):
    train_op = tf.train.RMSPropOptimizer(learning_rate).minimize(total_loss)

**Comment by VincentVon** [comment]
@hgffly , thank u for your advice!
I have change the corresponding codes at function "train()" in facenet.py, and using parameter "RMSPROP", but the problem still exists.
Do I changed codes at the right place? 
Thanks again!

**Comment by VincentVon** [comment]
@Julyustc , 
I have solved this issue! Advice of @hgffly gives a lot help.
First, use RMS optimizer, and this is most important;
Then, enlarge the epoch size, btw, I used 2000 and sucessed;
At last, I changed the center loss factor to 0.5.
Maybe these advices can help u.

**Comment by JingzheLyp** [comment]
Hello, are you solving this problem now? I have the same problem even I have used the idea of @hgffly 
 and @VincentVon , if it is resolved, can you share the solution about that?Thank you very much!

**Comment by VincentVon** [comment]
@JingzheLyp i did solved this problem. it seems the wrong parameters lead to this. my advice is changing the parameters mentioned before and changing optimizer. good luck.

**Comment by JingzheLyp** [comment]
Hi, ＠VincentVon, I have used your mobilenet code, and changed the center loss factor to 0.5, epoch size 2000, and in train_softmax.py, I changed the code as below：

     # Build a Graph that trains the model with one batch of examples and updates the model parameters
    # train_op = facenet.train(total_loss, global_step, args.optimizer, 
         #learning_rate, args.moving_average_decay, tf.global_variables(), args.log_histograms)    
     with tf.control_dependencies(tf.get_collection(tf.GraphKeys.UPDATE_OPS)):
            train_op =tf.train.RMSPropOptimizer(learning_rate).minimize(total_loss,global_step=global_step)

And parameters are：

max_nrof_epochs: 80
lfw_pairs: data/pairs.txt
embedding_size: 256
learning_rate_decay_factor: 1.0
center_loss_alfa: 0.9
center_loss_factor: 0.5
image_size: 224
weight_decay: 5e-05
keep_probability: 0.８

seed: 666
pretrained_model: None
log_histograms: True
learning_rate_decay_epochs: 100
nrof_preprocess_threads: 4

epoch_size: 2000
optimizer: RMSPROP
batch_size: 40
lfw_batch_size: 100
lfw_nrof_folds: 10
moving_average_decay: 0.9999
learning_rate: -1.0
gpu_memory_fraction: 0.9

random_flip: True
random_rotate: False
lfw_file_ext: png
filter_percentile: 100.0
filter_min_nrof_images_per_class: 0
random_crop: True

But the result is still
0	0.50000	1.00000
2000	0.50000	1.00000
4000	0.50000	1.00000
......

I have changed center_loss_factor, optimizer and weight_decay many times and got no good results, Can you give me more details about your training? Thank you very much!

**Comment by VincentVon** [comment]
@JingzheLyp hello, this is my parameters.

    max_nrof_epochs = 100
    batch_size = 32
    image_size = 224
    epoch_size = 2000

    embedding_size = 128
    random_crop = True
    random_flip = True
    random_rotate = True

    # keep_probability = 1.0 # 0.8
    # weight_decay = 0.0 # 5e-5
    # center_loss_factor = 0.0 # 1e-2
    # center_loss_alfa = 0.95 # 0.9
    learning_rate = 0.1  # -1

    keep_probability = 0.8
    weight_decay = 5e-5
    # center_loss_factor = 1e-2
    center_loss_factor = 0.5
    center_loss_alfa = 0.9
    # learning_rate = -1

    # optimizer = 'ADAGRAD'
    optimizer = 'RMSPROP'
    learning_rate_decay_epochs = 100
    learning_rate_decay_factor = 1.0
    moving_average_decay = 0.9999
    seed = 666
    nrof_preprocess_threads = 4
    log_histograms = True

    learning_rate_schedule_file = None
    filter_filename = ''
    filter_percentile = 100.0
    filter_min_nrof_images_per_class = 0

    # Parameters for validation on LFW
    lfw_file_ext = 'png'
    lfw_batch_size = 100
    lfw_nrof_folds = 10

**Comment by JingzheLyp** [comment]
Thank you,＠VincentVon, I want to know if I have changed the code at the right place? In train_softmax.py or facenet.py? if you change the code in  facenet.py, can you show me the changed part of the code? Thanks again!

**Comment by VincentVon** [comment]
@JingzheLyp hi, i have changed the codes from the original codes. My advice for u is changing parse_arguments in train_softmax.py. BTW, i think u should read original codes carefully. The original codes are well written and organized. Then u will know how to change.

**Comment by JingzheLyp** [comment]
＠VincentVon, hi, I just touched tensorflow for not long time, and i am newer to python. I have read the code several times and changed some parts, but I failed to make it, the result is still 
0	0.50000	1.00000
2000	0.50000	1.00000
4000	0.50000	1.00000
6000	0.50000	1.00000
8000	0.50000	1.00000
10000	0.50000	1.00000
12000	0.50000	1.00000
14000	0.50000	1.00000
16000	0.50000	1.00000
18000	0.50000	1.00000
20000	0.50000	1.00000
......
I wonder whether the hardware (GPU) leads to this problem. My GPU is GTX1060  6G memories. I really need you help, would you mind to provide me your codes? Thanks!

**Comment by Julyustc** [comment]
@VincentVon  hi, thanks to your help, I have solved the problem and got  96.7% accuracy. And I want to run the model on mobile device ,but I failed.  Have you tried? Thanks!

**Comment by tslater** [comment]
@Julyustc, I'm Impressed by your accuracy. Is that on LFW faces? Did you use the same parameters? What dataset did you train on?

**Comment by JingzheLyp** [comment]
@Julyustc hi, I have realized MTCNN on my mobile phone, but I have not solved the problem of training with mobilenet, can you shou your parameters and where you have changed your code?

**Comment by VincentVon** [comment]
@Julyustc well, i have trained mobilenet, but i have not use it on mobile device. i think its technical reason? what mobile device are u using?

**Comment by Julyustc** [comment]
@tslater  I used the same parameters as @VincentVon and trained on VGGFace2 dataset .

**Comment by Julyustc** [comment]
@VincentVon I used a Huawei phone. I want to implement face recognition on mobile device, but i dont't know how to load the model to the phone.

**Comment by VincentVon** [comment]
@Julyustc 华为手机……应该是国人吧，不知道tensorflow怎么直接部署到安卓上，可以考虑用下caffe，有教程部署在安卓手机上的。

**Comment by qwangku** [comment]
I came across the same problem (accuracy staying 0.5000+-0.000 after every epoch). What might be the potential cause for this issue?  At first, I thought something must be completely wrong, because it is not 0.4979 or 0.5021, but exactly 0.5000+-0.000, which looks like the network is doing nothing.
But I guess if embeddings of different faces are really similar, and the "threshold" is not right, it is very likely that different embeddings are all on the same side of the "threshold".

So can I say "accuracy stuck at 0.5000" does not indicate a fundamental fatal mistake (i.e. a dead network doing nothing)? Or it is very normal and just indicates the model is not good enough?

**Comment by abbyDC** [comment]
@VincentVon, @Julyustc  in what file (facenet.py or train_softmax.py) and what lines did you replace with the following code coming from @hgffly? 
`with tf.control_dependencies(tf.get_collection(tf.GraphKeys.UPDATE_OPS)):
train_op = tf.train.RMSPropOptimizer(learning_rate).minimize(total_loss)`

I still seem to have problems when I just change the parameters so maybe this part would solve it. Thanks in advance!

**Comment by tslater** [comment]
@Julyustc, do you ming sharing the parameters you used?
